File: src/com/castlefrog/agl/domains/backgammon/BackgammonState.java
package com.castlefrog.agl.domains.backgammon;

import com.castlefrog.agl.State;

/**
 * Represents a backgammon state as an array of byte locations. Each location is
 * 0 if no pieces are at that location and positive if player 1 has pieces there
 * and negative for the number of pieces player 2 has there.
 */
public final class BackgammonState implements State<BackgammonState> {
    private static final int NUMBER_OF_DICE = 2;
    private static final int NUMBER_OF_DIE_FACES = 6;
    private static final int NUMBER_OF_LOCATIONS = 26;

    private byte[] locations_;
    private byte[] dice_;
    private int agentTurn_;

    public BackgammonState() {
        locations_ = new byte[] {0, 2, 0, 0, 0, 0, -5, 0, -3, 0, 0, 0, 5, -5, 0, 0, 0, 3, 0, 5, 0, 0, 0, 0, -2, 0};
        dice_ = new byte[BackgammonState.getNumberOfDice()];

        do {
            dice_[0] = (byte) (Math.random() * BackgammonState.getNumberOfDieFaces() + 1);
            dice_[1] = (byte) (Math.random() * BackgammonState.getNumberOfDieFaces() + 1);
        } while (dice_[0] == dice_[1]);

        if (dice_[0] > dice_[1]) {
            dice_[1] = (byte) (Math.random() * BackgammonState.getNumberOfDieFaces() + 1);
            agentTurn_ = 0;
        } else {
            dice_[0] = (byte) (Math.random() * BackgammonState.getNumberOfDieFaces() + 1);
            agentTurn_ = 1;
        }
    }

    public BackgammonState(byte[] locations, byte[] dice, int agentTurn) {
        locations_ = locations;
        dice_ = dice;
        agentTurn_ = agentTurn;
    }

    public BackgammonState copy() {
        return this;
    }

    public byte[] getLocations() {
        byte[] locations = new byte[NUMBER_OF_LOCATIONS];
        for (int i = 0; i < NUMBER_OF_LOCATIONS; i += 1) {
            locations[i] = locations_[i];
        }
        return locations;
    }

    public byte getLocation(int index) {
        return locations_[index];
    }

    public byte[] getDice() {
        byte[] dice = new byte[NUMBER_OF_DICE];
        for (int i = 0; i < NUMBER_OF_DICE; i += 1) {
            dice[i] = dice_[i];
        }
        return dice;
    }

    public byte getDie(int number) {
        return dice_[number];
    }

    public static int getNumberOfDice() {
        return NUMBER_OF_DICE;
    }

    public static int getNumberOfDieFaces() {
        return NUMBER_OF_DIE_FACES;
    }

    public static int getNumberOfLocations() {
        return NUMBER_OF_LOCATIONS;
    }

    public int getAgentTurn() {
        return agentTurn_;
    }

    @Override
    public int hashCode() {
        int code = 11 * (7 + dice_[0]) + dice_[1];
        for (int i = 0; i < NUMBER_OF_LOCATIONS; i += 1) {
            code = 11 * code + locations_[i];
        }
        return code;
    }

    @Override
    public boolean equals(Object object) {
        if (!(object instanceof BackgammonState)) {
            return false;
        }
        BackgammonState state = (BackgammonState) object;
        for (int i = 0; i < NUMBER_OF_LOCATIONS; i += 1) {
            if (locations_[i] != state.getLocation(i)) {
                return false;
            }
        }
        return dice_[0] == state.getDie(0) && dice_[1] == state.getDie(1);
    }

    @Override
    public String toString() {
        StringBuilder output = new StringBuilder();
        output.append("[").append(dice_[0]).append("][").append(dice_[1]).append("]\n");
        for (int i = 12; i > 6; i--) {
            if (locations_[i] >= 0) {
                output.append(" ");
            }
            output.append(locations_[i]);
        }
        output.append("|");
        for (int i = 6; i > 0; i--) {
            if (locations_[i] >= 0) {
                output.append(" ");
            }
            output.append(locations_[i]);
        }
        output.append(" [").append(locations_[0]).append("]\n");
        output.append("------------|------------\n");
        for (int i = 13; i < 19; i++) {
            if (locations_[i] >= 0) {
                output.append(" ");
            }
            output.append(locations_[i]);
        }
        output.append("|");
        for (int i = 19; i < 25; i += 1) {
            if (locations_[i] >= 0) {
                output.append(" ");
            }
            output.append(locations_[i]);
        }
        output.append(" [").append(locations_[25]).append("]");
        return output.toString();
    }
}

File: src/main/kotlin/com/castlefrog/agl/domains/backgammon/BackgammonState.kt
package com.castlefrog.agl.domains.backgammon

import com.castlefrog.agl.State
import java.util.Arrays

/**
 * Represents a backgammon state as an array of byte locations.
 * @param locations each location is 0 if no pieces are at that location and positive for
 *                  the number of pieces player 1 has there and negative for the number of
 *                  pieces player 2 has there.
 * @param dice two values in range 0-5 that represent die faces 1 to 6 each; order does not matter
 * @param agentTurn current players turn
 */
class BackgammonState(
        val locations: ByteArray = byteArrayOf(0, 2, 0, 0, 0, 0, -5, 0, -3, 0, 0, 0, 5, -5, 0, 0, 0, 3, 0, 5, 0, 0, 0, 0, -2, 0),
        val dice: ByteArray = ByteArray(BackgammonState.N_DICE),
        var agentTurn: Int = 0) : State<BackgammonState> {

    override fun copy(): BackgammonState {
        val locations = ByteArray(N_LOCATIONS)
        System.arraycopy(this.locations, 0, locations, 0, N_LOCATIONS)
        val dice = ByteArray(N_DICE)
        System.arraycopy(this.dice, 0, dice, 0, N_DICE)
        return BackgammonState(locations, dice, agentTurn)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }
        if (other !is BackgammonState) {
            return false
        }
        return Arrays.equals(locations, other.locations) &&
                ((dice[0] == other.dice[0] && dice[1] == other.dice[1]) ||
                        (dice[0] == other.dice[1] && dice[1] == other.dice[0])) &&
                agentTurn == other.agentTurn
    }

    override fun hashCode(): Int {
        val dice = if (dice[0] < dice[1]) {
            intArrayOf(dice[1].toInt(), dice[0].toInt())
        } else {
            intArrayOf(dice[0].toInt(), dice[1].toInt())
        }
        return (agentTurn * 11 + Arrays.hashCode(locations)) * 17 + Arrays.hashCode(dice)
    }

    override fun toString(): String {
        val output = StringBuilder(" ").append(agentTurn).append(" - ")
        val dice = if (dice[0] < dice[1]) {
            intArrayOf(dice[1].toInt(), dice[0].toInt())
        } else {
            intArrayOf(dice[0].toInt(), dice[1].toInt())
        }
        output.append("[").append(dice[0] + 1).append("][").append(dice[1] + 1).append("]\n")
        for (i in 12 downTo 7) {
            if (locations[i] >= 0) {
                output.append(" ")
            }
            output.append(locations[i].toInt())
        }
        output.append("|")
        for (i in 6 downTo 1) {
            if (locations[i] >= 0) {
                output.append(" ")
            }
            output.append(locations[i].toInt())
        }
        output.append(" [").append(locations[0].toInt()).append("]\n")
        output.append("------------|------------\n")
        for (i in 13..18) {
            if (locations[i] >= 0) {
                output.append(" ")
            }
            output.append(locations[i].toInt())
        }
        output.append("|")
        var i = 19
        while (i < 25) {
            if (locations[i] >= 0) {
                output.append(" ")
            }
            output.append(locations[i].toInt())
            i += 1
        }
        output.append(" [").append(locations[25].toInt()).append("]")
        return output.toString()
    }

    companion object {
        val N_DICE = 2
        val N_DIE_FACES = 6
        val N_LOCATIONS = 26

        val TURN_PLAYER_1 = 0
        val TURN_PLAYER_2 = 1
    }
}
